fix(help): display error message if no option provided on startup

sumit-gupta91 opened the pull request:
added demandCommand options for yargs
by default requires one option to be speicified
will throw an error otherwise

<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**

**If relevant, did you update the documentation?**

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
#124 
**Does this PR introduce a breaking change?**
No

**Other information**

sumit-gupta91 commented:
Thanks for the review, will go through the tests and add the test case ASAP.

ev1stensberg commented:
Great! Ping me if you need help :)

ev1stensberg commented:
Seems like the bintestcases aren't running, can you use yarn install to see if that fixes the issue?

sumit-gupta91 commented:
I tried what was mentioned on webpack documentation.

`yarn install && yarn test` this throws and error as follows
`error  "webpack-dev-server" is not published  node/no-unpublished-require`

Yarn install works fine.
The only way i can run tests is `yarn jest`. This works fine but with the new changes I have added, some other tests are failing, shall I fix those test cases also.

ev1stensberg commented:
@sumit-gupta91 right, could you rebase against master? This should be fixed in a newer version of the master branch

ev1stensberg commented:
Aaah right, so we have 0CJS, you should add this at the very end of the processing of options, so the compiler can do a trial run. When webpack is called we have a try catch clause, could you add this in the catch block?

ev1stensberg commented:
HI @sumit-gupta91 , everything okay here? 💙

sumit-gupta91 commented:
Sorry guys for such a late response as I was on a leave for sometime, Wanted a break from work. I was not able to understand the last comment 

```
Aaah right, so we have 0CJS, you should add this at the very end of the processing of options, so the compiler can do a trial run. When webpack is called we have a try catch clause, could you add this in the catch block?
```

ev1stensberg commented:
So the reason why your test are failing is because webpack can compile without args, try adding the error message in the catch clause in bin/webpack

sumit-gupta91 commented:
I am not able to understand it, should I change the approach and rather than using demand I should use an if else in /bin/webpack.js or an enhancement has to be done. If it is the first option that has to be implemented I would like to understand the code flow.

ev1stensberg commented:
I think that you should let webpack try to parse without args, and if it output's any errors regarding no path specified, you can throw the help flag. If you look at the 0CJS PR, you can also see if there's anything missing from that in order to decide if we should output the help flag

sumit-gupta91 commented:
Okay will try this solution and update the PR

sumit-gupta91 commented:
Hi @ev1stensberg , I have updated the PR.

There are three tests which are failing, on the tests part I have two questions
a) How to run an individual test case in binTestCases folder ?
b) How to get the context on three test cases which are failing, what do the tests do, what real world scenario it is trying to replicate ?

Even though I have added the solution, it seems more of a hack to me. I still feel, if no options are specified webpack-cli should throw the error rather than webpack.

ev1stensberg commented:
> a) How to run an individual test case in binTestCases folder ?

Currently you can't

> b) How to get the context on three test cases which are failing, what do the tests do, what real world scenario it is trying to replicate ?

It's in the tests, read the file name and try to run the test input locally first ( like run the config with options in the test folder )

sendilkumarn commented:
> a) How to run an individual test case in binTestCases folder ?

actually you can hack a way into folder level. If you really like to add an if condition to check the folder name below this [line](https://github.com/webpack/webpack-cli/blob/d5c287bee55ca5ed70e9caa944114e9f80a4c57f/test/BinTestCases.test.js#L72)

ematipico commented:
With Jest also it is possible to do it

ev1stensberg commented:
Hey @sumit-gupta91 how's it going? Do you need any assistance?

sumit-gupta91 commented:
sorry guys was not able to look into it for sometime, doing it right away

sumit-gupta91 commented:
Hi @ev1stensberg, I have fixed the failing test cases. Do I need to add a new test case for error message on no options as some tests already test it.

dhruvdutt commented on test/binCases/no-options/development/stdin.js:
```diff
@@ -3,11 +3,14 @@
-	expect(stdout[4]).toContain("");
-	expect(stdout[5]).toContain("ERROR in Entry module not found");
-	expect(stdout[6]).toContain("");
+	console.log('printing value of error', stdout[1]);
```
Can you clean this?

dhruvdutt commented on test/binCases/no-options/none/stdin.js:
```diff
@@ -3,14 +3,17 @@
-	expect(stdout[7]).toContain("You can also set it to \'none\'");
-	expect(stdout[8]).toContain("");
-	expect(stdout[9]).toContain("ERROR in Entry module not found");
+	console.log('printing value of error', stdout[1]);
```
This as well.

ev1stensberg commented:
yes. And it shouldn't throw if `src/entry` and the defaults are present

ev1stensberg commented:
Try running webpack in a project without any options. Add `src/index.js` to the project. It should compile. With this PR it will throw an error, right?

sumit-gupta91 commented:
Okay let me check.

ematipico commented:
```bash
mkdir src
touch ./src/index.js
webpack
```

It should work

sumit-gupta91 commented:
This current PR does not throws an error if we run webpack on and a project which has a source src/index.js file. I just checked.

mkdir src
touch ./src/index.js
path to/webpack-cli/bin/webpack.js

sumit-gupta91 commented:
Great thanks @ev1stensberg.

ev1stensberg commented on bin/webpack.js:
```diff
@@ -482,6 +484,19 @@ For more information, see https://webpack.js.org/api/cli/.`);
 					);
 				} else if (stats.hash !== lastHash) {
 					lastHash = stats.hash;
+					if (stats.compilation && stats.compilation.errors.length !== 0) {
```
`&& outputOptions.entry === undefined`

sumit-gupta91 replied:
@ev1stensberg I would like to get some context on this comment ```Left a comment. Also, @sokra do you handle default setting in webpack internals? I think the CLI should do that. Sean did a PR but can't find the changes did there.```. What I can make of it is this ```if no option is provided then webpack-cli should throw an error and webpack should not be called```. Is my understanding correct ?

sumit-gupta91 commented:
@ev1stensberg I would like to get some context on this comment ```Left a comment. Also, @sokra do you handle default setting in webpack internals? I think the CLI should do that. Sean did a PR but can't find the changes did there.```. What I can make of it is this ```if no option is provided then webpack-cli should throw an error and webpack should not be called```. Is my understanding correct ?

ev1stensberg commented:
Alright, so try running your branch with arguments provided. That's what the change request is for 👍 

<img width="573" alt="skjermbilde 2018-05-12 kl 01 46 08" src="https://user-images.githubusercontent.com/16735925/39951043-8760d522-5586-11e8-8916-a453a95889ae.png">

sumit-gupta91 commented:
Hi @ev1stensberg I have updated the PR with the requested changes, If there are any more changes do tell me.

ev1stensberg commented:
@sumit-gupta91 Output is triggered two times now per run. Also, it won't exit webpack if the defaults aren't found, it will continue to run

sumit-gupta91 commented:
Hi @ev1stensberg have fixed the test cases and updated the code base.

EugeneHlushko commented on test/binCases/no-options/development/stdin.js:
```diff
@@ -1,13 +1,9 @@
-	expect(stdout[5]).toContain("ERROR in Entry module not found");
-	expect(stdout[6]).toContain("");
+	expect(stdout[3]).toContain("Insufficient number of arguments provided");
+	expect(stdout[4]).toContain("Alternatively, run `webpack(-cli) --help` for usage info");
```
im afraid people will literally run `webpack(-cli) --help` :(

EugeneHlushko commented on bin/webpack.js:
```diff
@@ -482,6 +484,21 @@ For more information, see https://webpack.js.org/api/cli/.`);
+						}
+					} else {
+						const statsString = stats.toString(outputOptions);
+						if (statsString) stdout.write(statsString + "\n");
```
Sorry for being out of context here, but from reading the PR history i dont understand why we are printing out output options on a successful compilation?

ev1stensberg replied:
the else statement should be discarded. On a failed compilation it will throw if no options are set, or?

EugeneHlushko replied:
oh yes, now i got the idea, thanks man! 👍

sumit-gupta91 commented:
Hi @ev1stensberg, I have merged the next branch in my current branch. What are the two dependencies to be removed ?. One more thing I did a merge instead of rebase will it make a difference ?

sumit-gupta91 commented:
I am not able to a yarn install. Second I am getting this error ""global-modules-path" is not found. (node/no-missing-require)" can someone please help me out here

ematipico commented:
Ouch, doing the merge caused some issue as far as I can see :( sometimes it's better to apply a rebase in order to not drag the commits from other branches. 

I think the better solution is to create a fresh new branch **from next** and doing a cherry-pick of your commits. Let's see

```bash
git fetch upstream
git checkout https://github.com/sumit-gupta91/webpack-cli
git checkout next
git checkout -b bugfix/error-message
git cherry-pick 580395a281c6dda405d15025f72f4c4418ce7315
git cherry-pick 37cb6d81a3f2d144e427361c0e48f663464e1a12
git cherry-pick d43ab9f8a08935d60b7523f8efbec80db80cbb7a
git cherry-pick 5a040bae8e4172bbdc20d385c2defe869cada975
git push
```

Try this locally. And after that, while trying to create a PR, see if the only changes that you see are yours only. Please direct your PR to **next**

sumit-gupta91 commented:
Yes this was I was stuck from quite sometime because next and master have conflicts, will create a new branch from next and then just add the files which were changed. @ematipico one more question in this branch whenever i am trying to do `yarn install ` i get an error no lock file found, am i doing something wrong here ?

sumit-gupta91 commented:
I am still not able to figure it out, on checking out a branch form next. If I do yarn install I get an error, if I do npm install it works, but npm run test does not works. What am i doing wrong here ?

ematipico commented:
We don't use yarn but we use npm for our developmenta. Give it a try and see if you get any errors.

About the tests, what kind of errors you get? Do you run them inside and integrated console inside your IDE? If so, that might be the problem, it happens to me sometime.

sumit-gupta91 commented:
I used npm install, after that npm run test throws this error.
```
lerna info version 2.11.0
lerna ERR! test Errored while running script in '@webpack-cli/webpack-scaffold'
lerna ERR! execute callback with error
lerna ERR! Error: Command failed: npm run test
lerna ERR! FAIL __tests__/index.test.js
lerna ERR!   ● Test suite failed to run
lerna ERR!
lerna ERR!     Cannot find module 'jscodeshift' from 'index.js'
lerna ERR!
lerna ERR!     > 1 | const jscodeshift = require("jscodeshift");
lerna ERR!       2 |
lerna ERR!       3 | function createArrowFunction(value) {
lerna ERR!       4 | 	return `() => '${value}'`;
lerna ERR!
lerna ERR!       at Resolver.resolveModule (../../node_modules/jest-resolve/build/index.js:169:17)
lerna ERR!       at Object.<anonymous> (index.js:1:110)
lerna ERR!
lerna ERR! Test Suites: 1 failed, 1 total
lerna ERR! Tests:       0 total
lerna ERR! Snapshots:   0 total
lerna ERR! Time:        1.093s
lerna ERR! Ran all test suites.
lerna ERR! npm ERR! code ELIFECYCLE
lerna ERR! npm ERR! errno 1
lerna ERR! npm ERR! @webpack-cli/webpack-scaffold@0.0.4 test: `jest`
lerna ERR! npm ERR! Exit status 1
lerna ERR! npm ERR!
lerna ERR! npm ERR! Failed at the @webpack-cli/webpack-scaffold@0.0.4 test script.
lerna ERR! npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
lerna ERR! npm WARN Local package.json exists, but node_modules missing, did you mean to install?
lerna ERR!
lerna ERR! npm ERR! A complete log of this run can be found in:
lerna ERR! npm ERR!     /Users/sumitgupta/.npm/_logs/2018-05-26T12_56_04_456Z-debug.log
lerna ERR!
lerna ERR! > @webpack-cli/webpack-scaffold@0.0.4 test /Users/sumitgupta/Work/Experiments/webpack-cli/packages/webpack-scaffold
lerna ERR! > jest
lerna ERR!
lerna ERR!
lerna ERR!     at Promise.all.then.arr (/Users/sumitgupta/Work/Experiments/webpack-cli/node_modules/lerna/node_modules/execa/index.js:236:11)
lerna WARN complete Waiting for 1 child process to exit. CTRL-C to exit immediately.
lerna ERR! test Errored while running script in '@webpack-cli/utils'
{ Error: Command failed: npm run test
```

I am using normal terminal to run the test and not the terminal associated with VSCode

sumit-gupta91 commented:
@ev1stensberg I opened a new branch from next but not able to run tests on it.
